query_optimizer/rules/UnnestSubqueries.cpp

```diff
@@ -0,0 +1,720 @@
+                                           inner_attributes,
+                                           probe_join_attributes,
+                                           build_join_attributes,
+                                           non_hash_join_predicates);
+  return output;
+}
+
+L::LogicalPtr UnnestSubqueriesForNonRootLogical::applyToNode(
+    const L::LogicalPtr &input,
+    E::UnorderedAttributeSet *inner_attributes,
+    std::vector<E::AttributeReferencePtr> *probe_join_attributes,
+    std::vector<E::AttributeReferencePtr> *build_join_attributes,
+    std::vector<E::PredicatePtr> *non_hash_join_predicates) {
+  // First eliminate subqueries expressions in this node (that is, this node is the outer query).
+  L::LogicalPtr input_with_no_subqueries = eliminateNestedScalarQueries(input);
+  DCHECK(input_with_no_subqueries != nullptr);
+
+  // Next determine whether this node has any outer reference (that is, this node is the inner query).
+  const std::vector<E::ExpressionPtr> &input_expressions = input_with_no_subqueries->input_expressions();
+  std::vector<E::ExpressionPtr> new_input_expressions;
+  bool has_change = false;
+  for (const E::ExpressionPtr &input_expression : input_expressions) {
+    const E::ExpressionPtr new_input_expression =
+        eliminateOuterAttributeReference(input_expression,
+                                         inner_attributes,
+                                         probe_join_attributes,
+                                         build_join_attributes,
+                                         non_hash_join_predicates);
+    if (new_input_expression != nullptr) {
+      new_input_expressions.emplace_back(new_input_expression);
+    }
+    if (new_input_expression != input_expression && !has_change) {
+      has_change = true;
+    }
+  }
+
+  switch (input_with_no_subqueries->getLogicalType()) {
+    case L::LogicalType::kAggregate: {
+      if (!non_hash_join_predicates->empty()) {
+        THROW_SQL_ERROR()
+            << "Non-equality join predicate with an outer query must be above any aggregate";
+      }
+
+      const L::Aggregate *aggregate = static_cast<const L::Aggregate*>(input_with_no_subqueries.get());
+      DCHECK(!has_change);
+      if (!inner_attributes->empty()) {
+        std::vector<E::NamedExpressionPtr> group_expressions = aggregate->grouping_expressions();
+        group_expressions.insert(group_expressions.end(),
+                                 inner_attributes->begin(),
+                                 inner_attributes->end());
+        return L::Aggregate::Create(aggregate->input(),
+                                    group_expressions,
+                                    aggregate->aggregate_expressions());
+      }
+      return input_with_no_subqueries;
+    }
+    case L::LogicalType::kFilter: {
+      if (new_input_expressions.empty()) {
+        const L::Filter* filter = static_cast<const L::Filter*>(input_with_no_subqueries.get());
```

Comment:
What do you think of applying the pointer sigil style here?

`const L::Filter *filter ...`